jerry-core/ecma/builtin-objects/ecma-builtin-function-prototype.c

```diff
@@ -244,57 +244,76 @@ ecma_builtin_function_prototype_object_bind (ecma_object_t *this_arg_obj_p , /**
-    ext_function_p = (ecma_extended_object_t *) function_p;
-    ECMA_SET_INTERNAL_VALUE_POINTER (ext_function_p->u.bound_function.target_function,
-                                     this_arg_obj_p);
+    bound_func_p = (ecma_bound_function_t *) function_p;
+    ECMA_SET_NON_NULL_POINTER_TAG (bound_func_p->header.u.bound_function.target_function,
+                                   this_arg_obj_p,
+                                   0);
 
-    ext_function_p->u.bound_function.args_len_or_this = ECMA_VALUE_UNDEFINED;
+    bound_func_p->header.u.bound_function.args_len_or_this = ECMA_VALUE_UNDEFINED;
 
     if (arguments_number != 0)
     {
-      ext_function_p->u.bound_function.args_len_or_this = ecma_copy_value_if_not_object (arguments_list_p[0]);
+      bound_func_p->header.u.bound_function.args_len_or_this = ecma_copy_value_if_not_object (arguments_list_p[0]);
     }
   }
   else
   {
     JERRY_ASSERT (arguments_number > 0);
 
-    size_t obj_size = sizeof (ecma_extended_object_t) + (arguments_number * sizeof (ecma_value_t));
+    size_t obj_size = sizeof (ecma_bound_function_t) + (arguments_number * sizeof (ecma_value_t));
 
     function_p = ecma_create_object (prototype_obj_p,
                                      obj_size,
                                      ECMA_OBJECT_TYPE_BOUND_FUNCTION);
 
     /* 8. */
-    ext_function_p = (ecma_extended_object_t *) function_p;
-    ECMA_SET_INTERNAL_VALUE_POINTER (ext_function_p->u.bound_function.target_function,
-                                     this_arg_obj_p);
+    bound_func_p = (ecma_bound_function_t *) function_p;
+    ECMA_SET_NON_NULL_POINTER_TAG (bound_func_p->header.u.bound_function.target_function,
+                                   this_arg_obj_p,
+                                   0);
 
     /* NOTE: This solution provides temporary false data about the object's size
        but prevents GC from freeing it until it's not fully initialized. */
-    ext_function_p->u.bound_function.args_len_or_this = ECMA_VALUE_UNDEFINED;
-    ecma_value_t *args_p = (ecma_value_t *) (ext_function_p + 1);
+    bound_func_p->header.u.bound_function.args_len_or_this = ECMA_VALUE_UNDEFINED;
+    ecma_value_t *args_p = (ecma_value_t *) (bound_func_p + 1);
 
     for (ecma_length_t i = 0; i < arguments_number; i++)
     {
       *args_p++ = ecma_copy_value_if_not_object (arguments_list_p[i]);
     }
 
     ecma_value_t args_len_or_this = ecma_make_integer_value ((ecma_integer_value_t) arguments_number);
-    ext_function_p->u.bound_function.args_len_or_this = args_len_or_this;
+    bound_func_p->header.u.bound_function.args_len_or_this = args_len_or_this;
   }
 
 #if ENABLED (JERRY_ES2015)
+  ecma_integer_value_t len = 0;
+  ecma_value_t length_value = ecma_op_object_find_own (ecma_make_object_value (this_arg_obj_p),
+                                                       this_arg_obj_p,
+                                                       ecma_get_magic_string (LIT_MAGIC_STRING_LENGTH));
```

Comment:
`Function.prototype.bind` based on [es6 19.2.3.2](https://www.ecma-international.org/ecma-262/6.0/#sec-function.prototype.bind)
Currently, this patch gets the length property of target (this_arg_obj_p) when it is an ordinary object case only.
For the case of Proxy, can I use `ecma_proxy_object_get_own_property_descriptor` to get the length property?

Reply:
Good remark, currently [[GetOwnProperty]] and [[DefineOwnProperty]] are the remaining unimplemented Proxy features. However the functions are exists so feel free to use them, but they will throw unimplemented error whenever they called. Nevertheless it's a good opportunity to ensure that the error handling is correct.

Reply:
Thank you for the review.
I fixed it to use `ecma_op_object_get_own_property_descriptor` to handle Proxy and ordinary object both.